Whiskey/Tasks/Invoke-WhiskeyParallelTask.ps1

```diff
@@ -0,0 +1,176 @@
+
+                    Import-Module -Name $whiskeyModulePath
+                    $moduleRoot = $whiskeyModulePath | Split-Path
+
+                    . (Join-Path -Path $moduleRoot -ChildPath 'Functions\Use-CallerPreference.ps1' -Resolve)
+                    . (Join-Path -Path $moduleRoot -ChildPath 'Functions\New-WhiskeyContextObject.ps1' -Resolve)
+                    . (Join-Path -Path $moduleRoot -ChildPath 'Functions\New-WhiskeyBuildMetadataObject.ps1' -Resolve)
+                    . (Join-Path -Path $moduleRoot -ChildPath 'Functions\New-WhiskeyVersionObject.ps1' -Resolve)
+
+                    # The task context gets serialized/deserialized into this new job process. We need to
+                    # correctly deserialize it back to an actual `Whiskey.Context` object. 
+                    $buildInfo = New-WhiskeyBuildMetadataObject
+                    Sync-ObjectProperty -Source $originalContext.BuildMetada -Destination $buildInfo -Exclude @( 'BuildServer' )
+                    if( $originalContext.BuildMetadata.BuildServer )
+                    {
+                        $buildInfo.BuildServer = $originalContext.BuildMetadata.BuildServer
+                    }
+                
+                    $buildVersion = New-WhiskeyVersionObject
+                    Sync-ObjectProperty -Source $originalContext.Version -Destination $buildVersion -ExcludeProperty @( 'SemVer1', 'SemVer2', 'SemVer2NoBuildMetadata' )
+                    $buildVersion.SemVer1 = $originalContext.Version.SemVer1.ToString()
+                    $buildVersion.SemVer2 = $originalContext.Version.SemVer2.ToString()
+                    $buildVersion.SemVer2NoBuildMetadata = $originalContext.Version.SemVer2NoBuildMetadata.ToString()
+
+                    $context = New-WhiskeyContextObject
+                    Sync-ObjectProperty -Source $originalContext -Destination $context -ExcludeProperty @( 'BuildMetadata', 'Version' )
+
+                    $context.BuildMetadata = $buildInfo
+                    $context.Version = $buildVersion
+
+                    $tasks = $using:queue['Tasks']
+                    foreach( $task in $tasks )
+                    {
+                        $taskName = $task.Keys | Select-Object -First 1
+                        $taskParameter = $task[$taskName]
+                        Invoke-WhiskeyTask -TaskContext $context -Name $taskName -Parameter $taskParameter
+                    }
+                }
+                $job | 
+                    Add-Member -MemberType NoteProperty -Name 'QueueIndex' -Value $queueIdx -PassThru |
+                    Add-Member -MemberType NoteProperty -Name 'Completed' -Value $false
+                [void]$jobs.Add($job)
+        }
+
+        $lastNotice = (Get-Date).AddSeconds(-61)
+        while( $jobs | Where-Object { -not $_.Completed } )
+        {
+            foreach( $job in $jobs )
+            {
+                if( $job.Completed )

```

Comment:
This seems like a redundant check since the `while` loop's condition is already doing `$jobs | Where-Object { -not $_.Completed }`.

Reply:
The `while` loop is checking if all the jobs are completed. The `if( $job.Completed )` clause is checking if each individual job as completed since the `foreach` loops always loops through each job.

Reply:
Oh yes, in my head I was reading `while` as `Foreach-Object` for some reason.